kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-backend/src/main/java/org/kie/workbench/common/stunner/bpmn/backend/converters/customproperties/InitializedVariable.java

```
@@ -71,6 +71,24 @@ public static InitializedInputVariable inputOf(
         }
     }
 
+    public static InitializedInputVariable createCustomInput(String parentId, VariableDeclaration varDecl, String expression) {
+        String isExpressionPattern = "^#\\{(.+)}$";
+        String decodedExpression = decode(expression);
+        if (decodedExpression.matches(isExpressionPattern)) {
+            return new InputExpression(parentId, varDecl, decodedExpression);
+        } else {
+            return new InputConstant(parentId, varDecl, decodedExpression);
+        }
+    }
+
+    private static String decode(String text) {
+        try {
+            return URLDecoder.decode(text, "UTF-8");
+        } catch (UnsupportedEncodingException e) {
+            throw new IllegalArgumentException(text, e);
```

By doing this you're converting the "checked" exception `UnsupportedEncodingException` into an "unchecked" one of type `IllegalArgumentException`. This results a bit "dangerous" because the signature for the `decode` method does not provide any exception information, but the decode can still fail, right? By doing this, actually it'll just thow an `IllegalArgumentException` and maybe it appears on the UI, but it's not really useful for the end user.
I was maybe expecting this decode method throwing the exception, and the caller for it catching the exception too, this way you can show a validation error to the user.
But again, not sure if makes completely sense, it depends on how to use it....

---

This catch `UnsupportedEncodingException` since we have encoding as not changing literal it should work without issues.

---

but I mean, if that `IllegalArgumentException` is being thrown at some point... what wil happen to the end user? does the exception will end up on the UI? does the field keep working properly? I'm just concerned about there this exception will end up....

---

@hasys any comments here?

---

Process won't be saved and in server log you will see:
<details>
  <summary>Click to expand</summary>
  <pre>ERROR 24-03 23:23:41,319 (BPMNStandaloneDiagramServiceImpl.java:transform:159) 	 Error whilst converting BPMN Diagram to XML.
java.lang.IllegalArgumentException: dsaf
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.decode(InitializedVariable.java:90)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.createCustomInput(InitializedVariable.java:78)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.inputOf(InitializedVariable.java:68)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.ParsedAssignmentsInfo.lambda$createInitializedInputVariables$0(ParsedAssignmentsInfo.java:191)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.ParsedAssignmentsInfo.createInitializedInputVariables(ParsedAssignmentsInfo.java:196)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.properties.ActivityPropertyWriter.setAssignmentsInfo(ActivityPropertyWriter.java:66)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.tasks.TaskConverter.userTask(TaskConverter.java:150)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$when$4(NodeMatch.java:85)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch$Case.match(NodeMatch.java:140)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$apply$5(NodeMatch.java:113)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.LinkedList$LLSpliterator.tryAdvance(LinkedList.java:1253)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.apply(NodeMatch.java:115)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.tasks.TaskConverter.toFlowElement(TaskConverter.java:64)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$when$4(NodeMatch.java:85)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch$Case.match(NodeMatch.java:140)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$apply$5(NodeMatch.java:113)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.LinkedList$LLSpliterator.tryAdvance(LinkedList.java:1253)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.apply(NodeMatch.java:115)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.FlowElementConverter.toFlowElement(FlowElementConverter.java:55)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1675)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:497)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.ProcessConverterDelegate.convertChildNodes(ProcessConverterDelegate.java:71)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverter.convertProcess(RootProcessConverter.java:52)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.DefinitionsConverter.toDefinitions(DefinitionsConverter.java:54)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.BaseDirectDiagramMarshaller.marshall(BaseDirectDiagramMarshaller.java:123)
	at deployment.business-central.war//org.kie.workbench.common.stunner.standalone.backend.services.BPMNStandaloneDiagramServiceImpl.transform(BPMNStandaloneDiagramServiceImpl.java:157)
	at deployment.business-central.war//org.kie.workbench.common.stunner.standalone.backend.services.BPMNStandaloneDiagramServiceImpl$Proxy$_$$_WeldClientProxy.transform(Unknown Source)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at deployment.business-central.war//org.jboss.errai.bus.server.io.AbstractRPCMethodCallback.invokeMethodFromMessage(AbstractRPCMethodCallback.java:65)
	at deployment.business-central.war//org.jboss.errai.bus.server.io.ValueReplyRPCEndpointCallback.callback(ValueReplyRPCEndpointCallback.java:40)
	at deployment.business-central.war//org.jboss.errai.bus.server.io.RemoteServiceCallback.callback(RemoteServiceCallback.java:54)
	at deployment.business-central.war//org.jboss.errai.cdi.server.CDIExtensionPoints$2.callback(CDIExtensionPoints.java:448)
	at deployment.business-central.war//org.jboss.errai.bus.server.DeliveryPlan.deliver(DeliveryPlan.java:47)
	at deployment.business-central.war//org.jboss.errai.bus.server.ServerMessageBusImpl.sendGlobal(ServerMessageBusImpl.java:297)
	at deployment.business-central.war//org.jboss.errai.bus.server.SimpleDispatcher.dispatchGlobal(SimpleDispatcher.java:46)
	at deployment.business-central.war//org.jboss.errai.bus.server.service.ErraiServiceImpl.store(ErraiServiceImpl.java:96)
	at deployment.business-central.war//org.jboss.errai.bus.server.service.ErraiServiceImpl.store(ErraiServiceImpl.java:113)
	at deployment.business-central.war//org.jboss.errai.bus.server.servlet.DefaultBlockingServlet.doPost(DefaultBlockingServlet.java:144)
	at javax.servlet.api@1.0.0.Final//javax.servlet.http.HttpServlet.service(HttpServlet.java:706)
	at javax.servlet.api@1.0.0.Final//javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at deployment.business-central.war//org.kie.workbench.common.stunner.standalone.backend.CORSFilter.doFilter(CORSFilter.java:78)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.opentracing.contrib.opentracing-jaxrs2//io.opentracing.contrib.jaxrs2.server.SpanFinishingFilter.doFilter(SpanFinishingFilter.java:55)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.core@2.0.13.Final//io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.core@2.0.13.Final//io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.core@2.0.13.Final//io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.core@2.0.13.Final//io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.core@2.0.13.Final//io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow@14.0.0.Final//org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet@2.0.13.Final//io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.core@2.0.13.Final//io.undertow.server.Connectors.executeRootHandler(Connectors.java:360)
	at io.undertow.core@2.0.13.Final//io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads@2.3.2.Final//org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1985)
	at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1487)
	at org.jboss.threads@2.3.2.Final//org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1378)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.io.UnsupportedEncodingException: UTF-
	at java.base/java.net.URLDecoder.decode(URLDecoder.java:144)
	at deployment.business-central.war//org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.decode(InitializedVariable.java:88)
	... 121 more
</pre>
</details>

Also it will be caught by failing test:
<details>
  <summary>InitializedVariableTest#urlDecodeConstants</summary>
  <pre>java.lang.IllegalArgumentException: %3C%3C%3C%23%21%21%21%23%3E%3E%3E

	at org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.decode(InitializedVariable.java:90)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.createCustomInput(InitializedVariable.java:78)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.properties.InitializedVariableTest.urlDecodeConstants(InitializedVariableTest.java:53)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:230)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:58)
Caused by: java.io.UnsupportedEncodingException: UTF-
	at java.lang.StringCoding.decode(StringCoding.java:190)
	at java.lang.String.<init>(String.java:426)
	at java.net.URLDecoder.decode(URLDecoder.java:190)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.customproperties.InitializedVariable.decode(InitializedVariable.java:88)
	... 25 more</pre>
</details>

I think it safe enough, since to get this error you need to modify and recompile the code (no user will do it) and also change literal to incorrect value and ignore the failing test.

---

We can ignore this exception and return empty value, for an instance and in this case in worse case user will lose only value for this property, but it will be very hard to debug.

---

Anyway user can do nothing to fix incorrect literal, because to fix it you need to modify the code and rebuild business central. So I am not sure how useful fancy message will be for the end user if someone will change for any reason this literal to incorrect one.